### PR TITLE
Restructured coverage functions

### PR DESCRIPTION
**Pull request by qiyunzhu**
Hello @dhakim87 I have carefully read and tested your code, and modified it based on my understanding. The performance of the code has been significantly improved. For a small test run:

- Old: 1:01.56 elapsed, 620644 maxresident
- New: 0:49.26 elapsed, 372700 maxresident

Modifications include:
- Decided that ordinal mapper does not need coverage calculation, because itself has coordinate operations. This simplifies logic.
- Created a new mapper: `range_mapper`, which is slightly modified from `plain_mapper` by returning range information. The old `plain_mapper` remains the same. This ensures that when the user chooses not to report coverage, the program performance remains the same.
- Replaced `SortedRangeList` class with functions to save class instantiation and member access overhead.
- Changed autocompress frequency from 100k to 10k. I haven't thoroughly benchmarked this parameter though.
- Multiple minor tweaks to improve performance.
- Reformatted code such as variable names to ensure consistent coding style.
- Added unit tests. Now coverage remains 100%.
- Merged v0.1.3 code.

However, there is still big room for improving the performance, considering that the analysis only takes 20 sec if one chooses not to report coverage. I will explore further tweaks.

Will appreciate your feedback! Once this PR is merged into your repo, your PR to my repo will be automatically updated, and I will review accordingly.

### CODE REVIEW

File: woltka/coverage.py
Comment (dhakim87) [inline woltka/coverage.py:48]: What is the purpose of this line?  Is this an optimization?  Why not just directly call res.append in the rest of this method?

File: woltka/coverage.py
Comment (dhakim87) [inline woltka/coverage.py]: What is the meaning of [0]?  Is this empty coverage?  Wouldn't that be []?

File: woltka/coverage.py
Comment (dhakim87) [inline woltka/coverage.py]: I don't understand the [0] here.

File: woltka/coverage.py
Comment (dhakim87) [inline woltka/coverage.py:84]: This looks like bad practice.  It feels very wrong to mix data types within a list, especially in a non interleaved way.  It will make all for loops more complicated, as evidenced by the use of the [1:] everywhere.  The chunking is a memory optimization, you can just as easily use len(cover) rather than cover[0] if you want to know when to run the merge step.  If you're intent on triggered based on number of items added rather than number of items kept, that count should be kept separate from the list itself- either creating an object that contains the list and the count, or creating two separate covers maps, covers containing the actual range lists and cover_counts keeping track of how many items have been added.

File: woltka/coverage.py
Comment (dhakim87) [reply woltka/coverage.py]: I believe that this syntax will make a copy of the merge_ranges result list that would not be made by cover[:] = merge_ranges(cover[1:] + ranges), but I can't say for sure that python won't optimize that away

File: woltka/coverage.py
Comment (dhakim87) [inline woltka/coverage.py]: I believe this syntax would avoid the copy I complain about above.

File: woltka/tests/test_coverage.py
Comment (dhakim87) [inline woltka/tests/test_coverage.py:122]: Might be worth a try finally to ensure the tmp directory is cleaned up when the test fails after writing coverage but asserts.

File: woltka/coverage.py
Comment (dhakim87) [inline woltka/coverage.py:28]: Can my name be credited somewhere as the source of the algorithm used in this function?

**Review by dhakim87 — APPROVED** [review]
Comments added.  Generally approve the changes, though I'd want the merge algorithm to be credited to me since it is nearly identical to the zebra filter implementation and the creation of the new file hides that it is stemming from my contribution.  

I think storing [items_added, range, range, range...] is bad practice, the (named)tuple (items_added, [range, range, range]) would be better, but that isn't my call to make.  

Between the chunking of operations and the storage of closures to avoid function lookups, there seems to be an large focus on low level optimization.  That screams to me that this is written in the wrong programming language.  Will this all need to be ported to C/C++ at some point?

File: woltka/coverage.py
Comment (qiyunzhu) [reply woltka/coverage.py:48]: Yes this is an optimization. It stores a reference to the method call so that Python doesn't need to search the attribute space of the variable for the particular method every time.

File: woltka/coverage.py
Comment (qiyunzhu) [reply woltka/coverage.py:84]: I agree with your opinion as I also feel bad with this solution.

It is ugly but it is the fastest solution I could think of. In this context, performance is the most important thing. Initially I used two separate dictionaries -- one for ranges and the other for counter. However, the query of dictionary is inefficient when there are many keys in it, which is the case here. Therefore, that solution (independent queries of two dictionaries) turned out to be inefficient.

Your suggestion of `[items_added, [range, range, range]]` is a good idea. I just tested. Your solution and my current solution have very close performance (runtime and memory). Let me update the PR using this solution.

I once considered using `len(cover)`. However, the logic prohibiting it is that the ranges may still be very fragmental after merging. This is very common when the sequencing depth is low for some genomes. Therefore, the length of `cover` may never drop below a certain line, no matter we merge it or not. If we use `len(cover)` as the mechanism to trigger auto-merge, the threshold must not be too small, and we don't know how big it should be. The worst case is that the function will futily merge the ranges during every call and no merge actually happens.

File: woltka/tests/test_coverage.py
Comment (qiyunzhu) [reply woltka/tests/test_coverage.py:122]: I think that the `tearDown` method already ensures that. See: https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown

File: woltka/coverage.py
Comment (qiyunzhu) [reply woltka/coverage.py:28]: Sure! I appreciate your contribution! I have added the credit to the docstring of `coverage.py`. How does it read?

**Comment by qiyunzhu** [comment]
Hello @dhakim87 Thank you for your review! I have implemented the data structure you suggested, which is neater and retains the same performance. I have also credited you in the docstring. Please kindly review. Thanks!

There is no plan to port to C/C++. I also think that this will significantly boost the performance. However, I consider that I am not as familiar with C/C++, and likely will not be due to the time commitment to my responsibilities. Meanwhile, continuous development and maintenance is important for the life cycle of the software tool. Therefore, I tend to stick to Python.

There has been an ongoing discussion with several collaborators regarding improving the performance of Woltka. They are expertised in computer science, especially hardware acceleration.

**Comment by qiyunzhu** [comment]
Here are some benchmarks of the chunk size (autocompress step size):

| Chunk | Runtime | Memory |
| :--- | ---: | ---: |
| 1000 | 0:58.17 | 260764 |
| 2000 | 0:52.82 | 282048 |
| 5000 | 0:49.19 | 329532 |
| 10000 | 0:49.57 | 372804 |
| 20000 | 0:50.49 | 429624 |
| 50000 | 0:53.74 | 538292 |
| 100000 | 0:57.23 | 654132 |

Therefore, **5000** may be the best, but this is a small-scale test so I would keep the current value 10000 until there is more evidence.

I think that no constant value will work the best for all sample - subject pairs. The ideal solution would be an **adaptive gradient** method which dynamically adjusts the merging rate according to the numbers of ranges merged in the past several rounds.

**Comment by qiyunzhu** [comment]
@dhakim87 I have found another significant optimization: Instead of storing range data as a list of tuples: `[(start, end), (start, end), (start, end)...]`, I flattened this structure into `[start, end, start, end, start, end...]`. This saves the compute overhead and memory cost of creating many tuples. When the ranges need to be merged, the flat list will be zipped up in pairs.

This optimization **reduced memory to half** (373228 to 192644), meanwhile it also reduced runtime slightly (0:49.06 to 0:46.99). Note that these numbers are for the entire Woltka run.

I plan to wait until you merge this PR, then submit another PR for you to review. What do you think?

**Comment by dhakim87** [comment]
Okay, I'll merge this PR now.

I think an interleaved start end list is a fine optimization

**Pull request merged**
